### ISSUE TITLE
[Bug] docs with dollar sign ($) cannot be found

### ISSUE THREAD

**Issue by tekumara** [dbt-labs/dbt-core#4386]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

```
$ dbt docs generate
Running with dbt=0.21.0
Encountered an error:
Compilation Error in model customers (models/customers.sql)
  Documentation for 'model.jaffle_shop.customers' depends on doc '__$start_lsn'  which was not found
```

### Expected Behavior

Doc names containing $ are found.

### Steps To Reproduce

_models/docs.md_:
```
{% docs __$start_lsn %}
Log sequence number (LSN) associated with the commit transaction for the change.
{% enddocs %}
```

In _schema.yml_ reference the doc using `{{ doc("__$start_lsn") }}`

### Environment

```markdown
- OS: Darwin
- Python: 3.9.7
- dbt: 0.21.0
```

**Comment by tekumara** [comment]
_mainfest.json_ looks correct:

```
$ jq -C '.nodes."model.jaffle_shop.customers".columns."__$start_lsn"' target/manifest.json      
{
  "name": "__$start_lsn",
  "description": "Log sequence number (LSN) associated with the commit transaction for the change.",
  "meta": {},
  "data_type": null,
  "quote": null,
  "tags": []
}
```

**Comment by tekumara** [comment]
BlockContents thinks the `BlockTag` name is `__`:
```
BlockContents(file=SourceFile(path=FilePath(searched_path='models', relative_path='docs.md', modification_time=1638433805.7870214, project_root='/Users/tekuma/code/jaffle_shop'), checksum=FileHash(name='sha256', checksum='3af2ad83d151cc6a03427eaf71f55200d0cb31c960a3d44d9527f1468314cf16'), project_name='jaffle_shop', parse_file_type=<ParseFileType.Documentation: 'docs'>, contents='{% docs __$start_lsn %}\nLog sequence number (LSN) associated with the commit transaction for the change.\n{% enddocs %}\n{% docs __seqval %}\nSequence value used to order the row changes within a transaction.\n{% enddocs %}', nodes=[], docs=[], macros=[]), block=BlockTag('docs', '__'))
```

**Comment by jtcohen6** [comment]
@tekumara Thanks for opening. This is the same issue as https://github.com/dbt-labs/dbt-core/issues/4352, reported just the other day. I'm going to close this issue as a duplicate, to keep the conversation centralized — I'd invite you to weigh in over there!

**tl;dr** All our Jinja block names are limited to alphanumeric characters and underscores (`r'[A-Za-z][A-Za-z_0-9]*'`). I'm inclined to view this as a reasonable limitation, and move that we document it as such. If there's a specific reason why you need to use special characters (such as `$`), I'd be curious to hear it over in #4352.
